doc/lspconfig.txt

```diff
@@ -1,242 +1,659 @@
-4. Quickstart                                           *lspconfig-quickstart*
+SETUP                                                          *lspconfig-setup*
+
+The main entrypoint to enabling and customizing a language server is the
+`setup {}` call. Using the defaults for a server built-into lspconfig is
+simple:
+>
+    require'lspconfig'.clangd.setup{}
+<
+The available server names are listed in `./server.md` and match the server
+name in `config.SERVER_NAME` defined in each configuration's source file.
+
+The purpose of `setup{}` is to wrap the call to `vim.lsp.start_client`  which
+is built into neovim core in an autocommand that calls `start_client` or
+`vim.lsp.buf_attach_client()` depending on if the current file appears to be
+in a project that contains a server previously started by lspconfig. See
+|root_dir| for more details.
+
+The `setup{}` function takes a table which contains a superset of the keys
+listed in `:help vim.lsp.start_client()` with the following unique entries:
+
+- {root_dir}
 
+    `function(filename, bufnr)`
 
-1. Install a language server, e.g. [pyright](CONFIG.md#pyright)
+    Returns either a filepath (string) or nil. The language server will only
+    start if the function returns a filepath.
 
-    ```bash
-    npm i -g pyright
-    ```
-2. Add the language server setup to your init.vim. The server name must match those found in the table of contents in [CONFIG.md](CONFIG.md)
+    If a root directory (string) is returned which is unique from any
+    previously returned root_dir, a new server will be spawned with that
+    root directory. See |lspconfig-root-directory| for more details
 
-    ```lua
-    lua << EOF
-    require'lspconfig'.pyright.setup{}
-    EOF
-    ```
-3. Create a project, this project must contain a file matching the root directory trigger. See [Automatically launching language servers](#Automatically-launching-language-servers) for additional info.
+- {name}
 
-    ```bash
-    mkdir test_python_project
-    cd test_python_project
-    git init
-    touch main.py
-    ```
-4. Launch neovim, the language server will now be attached and providing diagnostics (see `:LspInfo`)
+    `string`
+
+    Defaults to the server's name (`clangd`, `pyright`, etc.).
+
+- {filetypes}
+
+    `list[string] | nil`
+
+    Set of filetypes on which to attempt to resolve {root_dir}.
+
+    May be empty or server may specify a default value.
+
+- {autostart}
+
+    `bool` (default: true)
+
+    Controls if the `FileType` autocommand that launches a language server is
+    enabled. If `false`, allows for deferring language servers until manually
+    launched with `:LspStart` (|lspconfig-commands|).
+
+- {on_new_config}
+
+    `function(new_config, new_root_dir)`
+
+    Function executed after a root directory is detected. This is used to
+    modify the server configuration (including `cmd` itself). Most commonly,
+    this is used to inject additional arguments into `cmd`.
```

Comment:
You need to add an explanation about this as well: https://github.com/neovim/nvim-lspconfig/blob/564490ba5c48314e4c194279fa72dba9874bc65a/lua/lspconfig/angularls.lua#L52.

Reply:
This isn't true, or at least, the way I wrote it originally was extremely misleading.

Reply:
I'm not sure about the original use-case, but for `powershell_es` at least, I have to define both `on_new_config` and `cmd`, otherwise it wouldn't work.